internal/controller/etcd/register.go

```diff
@@ -0,0 +1,147 @@
+	builder := ctrl.
+		NewControllerManagedBy(mgr).
+		Named(controllerName).
+		WithOptions(controller.Options{
+			MaxConcurrentReconciles: r.config.Workers,
+			RateLimiter:             workqueue.NewItemExponentialFailureRateLimiter(10*time.Millisecond, r.config.EtcdStatusSyncPeriod),
+		}).
+		For(&druidv1alpha1.Etcd{}).
+		WithEventFilter(r.buildPredicate())
+
+	return builder.Complete(r)
+}
+
+// buildPredicate returns a predicate that filters events that are relevant for the Etcd controller.
+// NOTE:
+// For all conditions the following is applicable:
+// 1. create and delete events are always reconciled irrespective of whether reconcile annotation is present or auto-reconcile has been enabled.
+// 2. generic events are never reconciled. If there is a need in future to react to generic events then this should be changed.
+// Conditions for reconciliation:
+// Scenario 1: {Auto-Reconcile: false, Reconcile-Annotation-Present: false, Spec-Updated: false/true, Status-Updated: false/true, update-event-reconciled: false}
+// Scenario 2: {Auto-Reconcile: false, Reconcile-Annotation-Present: true, Spec-Updated: false, Status-Updated: false, update-event-reconciled: true}
+// Scenario 3: {Auto-Reconcile: false, Reconcile-Annotation-Present: true, Spec-Updated: false, Status-Updated: true, update-event-reconciled: false}
+// Scenario 4: {Auto-Reconcile: false, Reconcile-Annotation-Present: true, Spec-Updated: true, Status-Updated: false, update-event-reconciled: true}
+// Scenario 5: {Auto-Reconcile: false, Reconcile-Annotation-Present: true, Spec-Updated: true, Status-Updated: true, update-event-reconciled: true}
+// Scenario 6: {Auto-Reconcile: true, Reconcile-Annotation-Present: false, Spec-Updated: false, Status-Updated: false, update-event-reconciled: NA}, This condition cannot happen. In case of a controller restart there will only be a CreateEvent.
+// Scenario 7: {Auto-Reconcile: true, Reconcile-Annotation-Present: false, Spec-Updated: true, Status-Updated: false, update-event-reconciled: true}
+// Scenario 8: {Auto-Reconcile: true, Reconcile-Annotation-Present: false, Spec-Updated: false, Status-Updated: true, update-event-reconciled: false}
+// Scenario 9: {Auto-Reconcile: true, Reconcile-Annotation-Present: false, Spec-Updated: true, Status-Updated: true, update-event-reconciled: true}
+// Scenario 10: {Auto-Reconcile: true, Reconcile-Annotation-Present: false, Spec-Updated: false, Status-Updated: false, update-event-reconciled: false}
+// Scenario 11: {Auto-Reconcile: true, Reconcile-Annotation-Present: true, Spec-Updated: false, Status-Updated: false, update-event-reconciled: true}
+// Scenario 12: {Auto-Reconcile: true, Reconcile-Annotation-Present: true, Spec-Updated: true, Status-Updated: false, update-event-reconciled: true}
+// Scenario 13: {Auto-Reconcile: true, Reconcile-Annotation-Present: true, Spec-Updated: false, Status-Updated: true, update-event-reconciled: false}
```

Comment:
The same reasoning as above. In scenario 12, the spec is updated, and hence an event should be triggered. Whereas in scenario 13, spec is not updated, and only status is updated. So there should not be a new event triggered for it.

Reply:
make sense for scenario 12 and 13 but I'm more confused about scenario 11 when neither spec is updated nor status is updated then why new event is triggered for reconciliation ?

Reply:
Is this because of `Reconcile-Annotation-Present: true` ?

Reply:
Shreyas and I discussed this and we decided that as long as the reconcile annotation is present, reconciliation of the spec will be allowed. The reason is that at the end of the reconcile flow the annotation is removed. If there is a failure in one of the reconcile-spec flow then it will get re-queued automatically. Reconcile annotation will only be removed as the last step in the reconcile-spec flow.